phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -89,6 +94,319 @@
+        }
+
+        private static boolean isBeforeRebuildMissingIndexRowCount(Cell cell) {
+            if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                    IndexTool.BEFORE_REBUILD_MISSING_INDEX_ROW_COUNT_BYTES, 0,
+                    IndexTool.BEFORE_REBUILD_MISSING_INDEX_ROW_COUNT_BYTES.length) == 0) {
+                return true;
+            }
+            return false;
+        }
+
+        private static boolean isBeforeRebuildInvalidIndexRowCount(Cell cell) {
+            if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                    IndexTool.BEFORE_REBUILD_INVALID_INDEX_ROW_COUNT_BYTES, 0,
+                    IndexTool.BEFORE_REBUILD_INVALID_INDEX_ROW_COUNT_BYTES.length) == 0) {
+                return true;
+            }
+            return false;
+        }
+
+        private static boolean isAfterRebuildValidIndexRowCount(Cell cell) {
+            if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                    IndexTool.AFTER_REBUILD_VALID_INDEX_ROW_COUNT_BYTES, 0,
+                    IndexTool.AFTER_REBUILD_VALID_INDEX_ROW_COUNT_BYTES.length) == 0) {
+                return true;
+            }
+            return false;
+        }
+
+        private static boolean isAfterRebuildExpiredIndexRowCount(Cell cell) {
+            if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                    IndexTool.AFTER_REBUILD_EXPIRED_INDEX_ROW_COUNT_BYTES, 0,
+                    IndexTool.AFTER_REBUILD_EXPIRED_INDEX_ROW_COUNT_BYTES.length) == 0) {
+                return true;
+            }
+            return false;
+        }
+
+        private static boolean isAfterRebuildMissingIndexRowCount(Cell cell) {
+            if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                    IndexTool.AFTER_REBUILD_MISSING_INDEX_ROW_COUNT_BYTES, 0,
+                    IndexTool.AFTER_REBUILD_MISSING_INDEX_ROW_COUNT_BYTES.length) == 0) {
+                return true;
+            }
+            return false;
+        }
+
+        private static boolean isAfterRebuildInvalidIndexRowCount(Cell cell) {
+            if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                    IndexTool.AFTER_REBUILD_INVALID_INDEX_ROW_COUNT_BYTES, 0,
+                    IndexTool.AFTER_REBUILD_INVALID_INDEX_ROW_COUNT_BYTES.length) == 0) {
+                return true;
+            }
+            return false;
+        }
+
+        private long getValue(Cell cell) {
+            return Long.parseLong(Bytes.toString(cell.getValueArray(),
+                    cell.getValueOffset(), cell.getValueLength()));
+        }
+
+        private void update(Cell cell) {
+            if (isScannedDataRowCount(cell)) {
```

Comment:
if this method knew what phase it was in you could eliminate half the comparisons, right?

Reply:
This method is called only once and by the reducer. It is not called by the mapper which implements the phases.